keps/sig-node/20191025-image-pull-progress.md

```diff
@@ -0,0 +1,402 @@
+    string reason = 4;
+}
+```
+
+The `PullImageState` provides richer information to the kubelet in which state
+the image pull currently resides as well as a fine granular error messages
+directly from the runtimes via the `reason` field.
+
+### Kubernetes API additions
+
+The already existing kubelet `imageManager` events `PullingImage`, `PulledImage`
+and `FailedToPullImage` will be re-used and extended with the additional
+metadata provided by the CRI. This means that the container will still reside in
+`ContainerStateWaiting` during the image pull, but the events regarding the
+image pull process will be thrown periodically. The amount of how often the
+kubelet throws this event should be configurable in a range between 1s and 1min.
+
+To reduce the impact of the added feature, a kubelet configuration option will
+be added which is (for now) disabled per default. This option completely avoids
+calling the new gRPC API and falls back to the current `PullImage` RPC.
+
+### CLI modifications
+
+The command line interface of Kubernetes (`kubectl`) does right now provide all
+necessary features to display the additional pull progress events, but the
+overall usability is not satisfying at all.
+
+To improve the amount of information when calling `kubectl get pods`, we will
+now take the thrown events into account as well. To achieve that and stay
+backwards compatible, a new output variant `events` will be added to the
+`kubectl get` flag `-o/--output`. If the flag is set, `kubectl` will map
+(manually implemented) relevant information into the default (table) output in a
+similar fashion to the `-o wide`. For example, the output for pods pulling a
+container image will now look like:
+
+```
+> kubectl get pods -o events
+NAMESPACE     NAME                       READY   STATUS              RESTARTS   AGE   EVENT
+kube-system   coredns-65c865486d-s2fs5   0/1     ContainerCreating   0          3s    Pulling Image (42%)
+```
+
+### Implementation Details/Notes/Constraints
+
+The major caveat for this implementation approach is the increased amount of
+thrown events from the kubelet during an image pull. This should be limited in a
+fashion that the API surface is not strongly impacted.
+
+### Risks and Mitigations
+
+The interface changes to `kubectl` should increase the usability to the users,
+whereas it opens the doors for future correlations between events and
+`kubectl get` requests. There is a risk that the current design proposal does
+not fulfill that need.
+
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  Via the kubelet dynamic configuration feature, where we will add a dedicated
+  option for it.
+
+- **Can the feature be disabled once it has been enabled (i.e., can we roll
+  back the enablement)?**
+
+  Yes, but not during a image pull (already established gRPC stream).
```

Comment:
> but not during a image pull (already established gRPC stream)

Since kubelet got the `ctx`, it could still close the `ctx` when the feature got disabled.

Reply:
I agree, I will make that more clear.